### ISSUE TITLE
[BUG] Upgrade stuck due to alternation to the default StorageClass in a cluster containing a witness node

### ISSUE THREAD

**Issue by mattmattox** [harvester/harvester#7072]
---

**Describe the bug**  
If you edit the default storage class `harvester-longhorn`, the v1.4.0 upgrade will get stuck due to the following error.  
NOTE: I changed the replica count to 2 as this is a 2-node cluster with a witness node.  

```
REVISION	UPDATED                 	STATUS  	CHART          	APP VERSION	DESCRIPTION                                                                                                                                                                                                   
3       	Tue Nov 26 01:58:21 2024	deployed	harvester-1.3.2	v1.3.2     	Upgrade complete                                                                                                                                                                                              
117     	Sun Dec  1 06:24:55 2024	failed  	harvester-1.4.0	v1.4.0     	Upgrade "harvester" failed: cannot patch "harvester-longhorn" with kind StorageClass: StorageClass.storage.k8s.io "harvester-longhorn" is invalid: parameters: Forbidden: updates to parameters are forbidden.
```

---

**To Reproduce**  
1. Follow the steps at [[Harvester Witness Node Configuration](https://docs.harvesterhci.io/v1.3/advanced/witness/)](https://docs.harvesterhci.io/v1.3/advanced/witness/).  
2. Instead of creating a new storage class, edit the `harvester-longhorn` storage class directly and modify parameters (e.g., change replica count).  
3. Perform the upgrade to v1.4.0.  

---

**Expected behavior**  
The upgrade process should complete successfully even if the `harvester-longhorn` storage class has been modified, or there should be a clear pre-upgrade check or migration step to handle the modified storage class gracefully.

---

**Environment**  
- **Harvester ISO version**: 1.3.2  
- **Underlying Infrastructure**: Baremetal with Dell PowerEdge R420  

---

**Additional context**  
This issue highlights a potential gap in upgrade compatibility when users modify built-in resources like storage classes. If such modifications are unsupported, this should be documented, and the upgrade should preemptively fail with a descriptive message.

---

**Comment by starbops** [comment]
Hi @mattmattox, Thank you for reporting to us. Just want to confirm: did you remove/re-add or replace the default `harvester-longhorn` StorageClass with the current one whose `numberOfReplicas` is `2`?

Please attach a support bundle and upgrade log archive, if possible, so we can investigate further.

P.S. I updated the issue title to better describe the situation. If you find it's not accurate, please update it directly. Thank you.

**Comment by mattmattox** [comment]
> Hi @mattmattox, Thank you for reporting to us. Just want to confirm: did you remove/re-add or replace the default `harvester-longhorn` StorageClass with the current one whose `numberOfReplicas` is `2`?
> 
> Please attach a support bundle and upgrade log archive, if possible, so we can investigate further.

Yes, I grabbed a YAML output, deleted the storageclass, edited the YAML (set numberOfReplicas=2),  and reapplied.

I was able to work around this issue by deleting the storageclass and letting the helm chart put it back to the default.

[hvst-upgrade-kv57w-upgradelog-archive-2024-11-30T19-53-17Z.zip](https://github.com/user-attachments/files/17972299/hvst-upgrade-kv57w-upgradelog-archive-2024-11-30T19-53-17Z.zip)

**Comment by starbops** [comment]
Thanks for the confirmation. Ideally, the Upgrade validating webhook should've blocked the upgrade since the `harvester` ManagedChart is unready due to the alternation to the `harvester-longhorn` StorageClass. I wonder why this did not happen.

I will try to reproduce it in my environment. Please let us know if you have any other manual operations before the upgrade.

**Comment by Vicente-Cheng** [comment]
Thanks @mattmattox,

Yeah, as @starbops mentioned. Modifying the default StorageClass `harvester-longhorn` will make the upgrade fail because of incomplete `ManagedChart`. 

I thought we needed to block the parameter modification (especially the replica number) of the longhorn StorageClass because we should make consistency with the whole PVC on the same StorageClass.
e.g., Some volumes created from `harvester-longhorn` were replicas 3. Others were replicas 2. It should not happened.

WDYT? @ibrokethecloud, @bk201

**Comment by Yu-Jack** [comment]
I think we could add the webhook to block the request of modifying the default SC. After all, we already have [[ENHANCEMENT] Do not allow to delete harvester-longhorn storageclass](https://github.com/harvester/harvester/issues/6678). But, for those users who haven’t upgraded yet, they won't have this webhook, so we can add a warning message about this topic in [Upgrade Harvester](https://docs.harvesterhci.io/v1.4/upgrade/index).

**Comment by starbops** [comment]
If you try to modify the parameters in a StorageClass, it will be blocked:

```
storageclasses.storage.k8s.io "harvester-longhorn" was not valid: parameters: Forbidden: updates to parameters are forbidden.
```

Our validating webhook does not block it, but I don't find who does this. FWIW, there are two relevant issues in the upstream, both of them were not handled:
- kubernetes/kubernetes#80472
- kubernetes/kubernetes#106996

With that said, I cannot reproduce this in a standard 3-node cluster setup (the default SC is replaced with a new one with `numberOfReplicas=2`). The upgrade request will be blocked as expected due to harvester managed chart not being ready:

```
# harvester ManagedChart
  ...
  summary:
    desiredReady: 1
    modified: 1
    nonReadyResources:
    - bundleState: Modified
      modifiedStatus:
      - apiVersion: storage.k8s.io/v1
        kind: StorageClass
        name: harvester-longhorn
        patch: '{"parameters":{"numberOfReplicas":"3"}}'
      name: fleet-local/local
    ready: 0
    ...
```

<img width="849" alt="image" src="https://github.com/user-attachments/assets/47746a72-6bc4-477f-8148-09dfb6ea943c">

@Vicente-Cheng Do you have any insight about this? Will this be different in clusters with a witness node?

**Comment by bk201** [comment]
The suggested way is to create a new sc and make it default like: https://docs.harvesterhci.io/v1.3/advanced/witness/#limitations

**Comment by starbops** [comment]
Perhaps we can re-iterate and emphasize the need to keep the original default StorageClass, `harvester-longhorn`, despite the fact that its numberOfReplicas might not fit the user's cluster formation. The best practice is to create **another** new default StorageClass, not replacing the original one with different parameters. It's important from the Harvester upgrade's POV.

That being said, a potential bug might exist in the upgrade validating webhook. Given that the ManagedChart is unready, it's unusual for the webhook not to block such a request.

**Comment by w13915984028** [comment]
The issue is closed.
 
1. The error `storageclasses.storage.k8s.io "harvester-longhorn" was not valid: parameters: Forbidden: updates to parameters are forbidden.` comes from k8s internal, on the SC CRD definition, the `parameters` are immutable after it's creation.

@mattmattox Used a hack solution to skip: delete the original SC and create a modified one.

2. On Harvester v1.3.* cluster, if you encounter this issue, the workaround is to delete the `harvester-longhorn` and re-create it to recover to the default value https://github.com/harvester/harvester/issues/7072#issuecomment-2510523298

3. On Harvester v1.4.*, with https://github.com/harvester/harvester/issues/6678, the `harvester-longhorn` is not allowed to be deleted.


4. The suggested way is to create a new sc and make it default like: https://docs.harvesterhci.io/v1.3/advanced/witness/#limitations https://github.com/harvester/harvester/issues/7072#issuecomment-2518874429
